### PR TITLE
Add Windows (msbuild & cygwin) CI

### PR DESCRIPTION
**Pull request by z3ntu**
This PR also contains "Upgrade Windows projects to VS2015", originally signal11/hidapi#324 (but `hidtest.cpp` replaced with `hidtest.c`) as the old project file format is not compatible with Visual Studio 2015 (the Visual Studio version appveyor uses).

@Qbicz has changes for 64-bit builds, that should be integrated as well I think (and then the CI script changed to build both)

### CODE REVIEW

**Comment by z3ntu** [comment]
Probably adding `/p:Platform=%arch%` to the msbuild command will build 32/64-bit builds.

File: windows/hidapi.vcxproj [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+      <TargetMachine>MachineX86</TargetMachine>
+    </Link>
+  </ItemDefinitionGroup>
+  <ItemDefinitionGroup Condition="'$(Configuration)|$(Platform)'=='Release|Win32'">
+    <ClCompile>
+      <Optimization>MaxSpeed</Optimization>
+      <IntrinsicFunctions>true</IntrinsicFunctions>
+      <AdditionalIncludeDirectories>..\hidapi;%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
+      <PreprocessorDefinitions>WIN32;NDEBUG;_WINDOWS;_USRDLL;HIDAPI_EXPORTS;%(PreprocessorDefinitions)</PreprocessorDefinitions>
+      <RuntimeLibrary>MultiThreadedDLL</RuntimeLibrary>
```
Comment (Youw) [inline windows/hidapi.vcxproj]: this is where we can avoid dependencies on MSVC dll runtime:
```suggestion
      <RuntimeLibrary>MultiThreaded</RuntimeLibrary>
```

File: windows/hidtest.vcxproj [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+      <Message>Copying hidapi.dll to the local direcotry.</Message>
+      <Command />
+    </PostBuildEvent>
+  </ItemDefinitionGroup>
+  <ItemDefinitionGroup Condition="'$(Configuration)|$(Platform)'=='Release|Win32'">
+    <ClCompile>
+      <Optimization>MaxSpeed</Optimization>
+      <IntrinsicFunctions>true</IntrinsicFunctions>
+      <AdditionalIncludeDirectories>..\hidapi;%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
+      <RuntimeLibrary>MultiThreadedDLL</RuntimeLibrary>
```
Comment (Youw) [inline windows/hidtest.vcxproj]:
```suggestion
      <RuntimeLibrary>MultiThreaded</RuntimeLibrary>
```

**Comment by z3ntu** [comment]
@Qbicz I've cherry-picked https://github.com/Qbicz/hidapi/commit/a0a00ca76bbc6a2bad1b235f23c96fac6853c30b into this PR but msbuild complains `error MSB8013: This project doesn't contain the Configuration and Platform combination of Release|x64. [C:\projects\hidapi\windows\hidapi.vcxproj]` . Could you look into that?

**Comment by Youw** [comment]
I can check

**Comment by Youw** [comment]
@z3ntu - go ahead and cherry-pick https://github.com/libusb/hidapi/commit/bddd830ff3606a812e0a0e186e1528481fe50f1a

File: .appveyor.yml [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+build_script:
+  - cmd: if %BUILD_ENV%==msbuild (
+        msbuild .\windows\hidapi.sln /p:Configuration=Release /p:Platform=%arch% /logger:"C:\Program Files\AppVeyor\BuildAgent\Appveyor.MSBuildLogger.dll" )
+  - cmd: if %BUILD_ENV%==cygwin (
+        C:\cygwin64\bin\bash -exlc "cd $APPVEYOR_BUILD_FOLDER; ./bootstrap; ./configure; make" )
+
+artifacts:
+  - path: .\windows\Release\hidapi.dll
+  - path: .\windows\Release\hidapi.lib
+  - path: .\windows\Release\hidapi.pdb
```
Comment (Youw) [inline .appveyor.yml]: do we have `.pdb` files for `cygwin` build?
or is it silently ignores its absence?

Comment (z3ntu) [reply .appveyor.yml]: Non-existing artifacts get ignored by appveyor

**Comment by z3ntu** [comment]
Builds: https://ci.appveyor.com/project/z3ntu/hidapi/builds/25960104

But there are some warnings:

In Win32:
```
  C:\Program Files (x86)\MSBuild\Microsoft.Cpp\v4.0\V140\Microsoft.CppBuild.targets(392,5): warning MSB8028: The intermediate directory (Release\) contains files shared from another project (hidapi.vcxproj).  This can lead to incorrect clean and rebuild behavior. [C:\projects\hidapi\windows\hidtest.vcxproj]
```
In x64:
```
hid.c(650): warning C4267: 'function': conversion from 'size_t' to 'DWORD', possible loss of data [C:\projects\hidapi\windows\hidapi.vcxproj]
hid.c(693): warning C4267: 'function': conversion from 'size_t' to 'DWORD', possible loss of data [C:\projects\hidapi\windows\hidapi.vcxproj]
hid.c(747): warning C4267: 'return': conversion from 'size_t' to 'int', possible loss of data [C:\projects\hidapi\windows\hidapi.vcxproj]
hid.c(763): warning C4267: 'function': conversion from 'size_t' to 'ULONG', possible loss of data [C:\projects\hidapi\windows\hidapi.vcxproj]
hid.c(769): warning C4267: 'return': conversion from 'size_t' to 'int', possible loss of data [C:\projects\hidapi\windows\hidapi.vcxproj]
hid.c(791): warning C4267: 'function': conversion from 'size_t' to 'DWORD', possible loss of data [C:\projects\hidapi\windows\hidapi.vcxproj]
hid.c(792): warning C4267: 'function': conversion from 'size_t' to 'DWORD', possible loss of data [C:\projects\hidapi\windows\hidapi.vcxproj]
hid.c(833): warning C4267: 'function': conversion from 'size_t' to 'ULONG', possible loss of data [C:\projects\hidapi\windows\hidapi.vcxproj]
hid.c(846): warning C4267: 'function': conversion from 'size_t' to 'ULONG', possible loss of data [C:\projects\hidapi\windows\hidapi.vcxproj]
hid.c(859): warning C4267: 'function': conversion from 'size_t' to 'ULONG', possible loss of data [C:\projects\hidapi\windows\hidapi.vcxproj]
hid.c(872): warning C4267: 'function': conversion from 'size_t' to 'ULONG', possible loss of data [C:\projects\hidapi\windows\hidapi.vcxproj]
```
(and also the warning from Win32)

Also, is it normal that the arches are called "Win32" and "x64" and not either x86/x64 or Win32/Win64?

And the output folder from x64 builds is in a subdirectory while Win32 is not.

**Comment by Youw** [comment]
> Also, is it normal that the arches are called "Win32" and "x64" and not either x86/x64 or Win32/Win64?

`Win32` is the default name for the configuration name on older versions of Visual Studio (not sure since when, but surely it was even before x64 was (commonly) used).

Currently (MSVC2017 for sure, didn't check others), by default creates configurations `x86` and `x64`, so that is common for legacy projects to have exactly Win32/x64 pair of configurations.

> And the output folder from x64 builds is in a subdirectory while Win32 is not.

That is a default behavior of Visual Studio. It is configurable, but I'd use defaults as much as possible.

> C:\Program Files (x86)\MSBuild\Microsoft.Cpp\v4.0\V140\Microsoft.CppBuild.targets(392,5): warning MSB8028: The intermediate directory (Release\) contains files shared from another project (hidapi.vcxproj).  This can lead to incorrect clean and rebuild behavior. [C:\projects\hidapi\windows\hidtest.vcxproj]

warning makes perfect sense, even though there is nothing wrong in our case
we can play with project settings (artifacts location) to get rid of it, but to make it right properly, we need to move each project (hidapi/hidtest) into a separate folder - Visual Studio always creates a separate folder for each new project

don't thing it is a big deal right now (at least not in this PR)

> x64 warnings

best we can do (without runtime overhead) - explicitly silence them (explicitly cast); don't think it is critical either

---
my bigger concern right now - AppVeyor gathers only `MSVC Win32` artifacts
my opinion - mingw artifacts are optional, but `MSVC x64` is a must
I think it can be easily fixed, if you add a variable to a build environment matrix with artifacts path

**Comment by z3ntu** [comment]
> my bigger concern right now - AppVeyor gathers only MSVC Win32 artifacts

done, but it should be verified that the artifacts are really 32- and 64-bit


Okay so from the general changes this should be good to go, I'll just clean up the git commits a bit

**Comment by Qbicz** [comment]
Big thank you @z3ntu and @Youw for testing this. I currently don't have any Windows machine set up to test this.

I have seen these x64 warnings before. As suggested by @Youw we can cast these, but this is not critical.

Looks good to me.

**Comment by z3ntu** [comment]
Someone with permissions for the repo still needs to set up appveyor I think. 

Also as I'm using `.appveyor.yml` instead of `appveyor.yml` (note the dot), appveyor needs to be told to use that file under "Settings -> General -> Custom configuration .yml file name" on appveyor.com.

**Comment by Youw** [comment]
@LudovicRousseau I've sent a request on installing AppVeyor on `hidapi`, can you approve it?

**Comment by LudovicRousseau** [comment]
@Youw done

**Comment by Youw** [comment]
Thanks, checking

**Comment by Youw** [comment]
Hmm, so the build configuration is working, but I missed the fact, that AppVeyor Projects better to be created by the organisation, otherwise is url is user specific: https://ci.appveyor.com/project/Youw/hidapi

@LudovicRousseau, can you create a project as libusb organisation project?
To do so:
1) Sign in as organisation
2) Projects -> New Project -> GitHub -> libusb/hidapi

All projects settings are fine to be default except one:
image

**Comment by LudovicRousseau** [comment]
I will see that in August.
Ping me again if I forget.

**Comment by Youw** [comment]
Sure, np. Until then - we'll use my personal project configuration.
https://github.com/libusb/hidapi/issues/47 as a reminder.

**Pull request merged**